Yosemite/Yosemite/Tools/Stats/TopEarnerStatsCachingFetcher.swift

```diff
@@ -0,0 +1,90 @@
+import Foundation
+import Networking
+
+private struct ProductsReportRequestResponse: Codable {
+    let items: [ProductsReportItem]
+}
+
+class TopEarnerStatsCachingFetcher {
+    private let productsReportsRemote: ProductsReportsRemote
+    private let cache: CodablePersistentCache<ProductsReportRequestResponse>
+
+    init(network: Network) {
+        self.productsReportsRemote = ProductsReportsRemote(network: network)
+        self.cache = CodablePersistentCache<ProductsReportRequestResponse>()
+    }
+
+    @MainActor
+    func loadTopEarnerStats(siteID: Int64,
+                            timeRange: StatsTimeRangeV4,
+                            timeZone: TimeZone,
+                            earliestDateToInclude: Date,
+                            latestDateToInclude: Date,
+                            quantity: Int) async throws -> TopEarnerStats {
+        let cacheKey = requestCacheKey(siteID: siteID,
+                                       timeZone: timeZone,
+                                       earliestDateToInclude: earliestDateToInclude,
+                                       latestDateToInclude: latestDateToInclude,
+                                       quantity: quantity)
+        let response: [ProductsReportItem]
+        if let cachedResponse: ProductsReportRequestResponse = try? cache.load(forKey: cacheKey) {
+            response = cachedResponse.items
+        } else {
+            response = try await productsReportsRemote.loadTopProductsReport(for: siteID,
+                                                                                       timeZone: timeZone,
+                                                                                       earliestDateToInclude: earliestDateToInclude,
+                                                                                       latestDateToInclude: latestDateToInclude,
+                                                                                       quantity: quantity)
+            let cacheEntry = CodableCacheEntry(value: ProductsReportRequestResponse(items: response), timeToLive: Constants.cacheTimeToLive)
+            cache.save(cacheEntry, forKey: cacheKey)
```

Comment:
I decided to cache the `[ProductsReportItem]` instead of `TopEarnerStats` as the former is what we get remotely.  We generate `TopEarnerStats` out of `[ProductsReportItem]`, so caching the network response would allow us to skip several network requests (i.e. different granularity would have the same network requests, but different `TopEarnerStats`